compiler/rustc_mir_transform/src/check_alignment.rs

```diff
@@ -63,69 +64,89 @@ impl<'tcx> MirPass<'tcx> for CheckAlignment {
+                | MutatingUseContext::AsmOutput
+                | MutatingUseContext::Call
+                | MutatingUseContext::Yield
+                | MutatingUseContext::Drop
+                | MutatingUseContext::Borrow,
+            ) => {}
+            PlaceContext::NonMutatingUse(
+                NonMutatingUseContext::Inspect
+                | NonMutatingUseContext::Copy
+                | NonMutatingUseContext::Move
+                | NonMutatingUseContext::SharedBorrow
+                | NonMutatingUseContext::ShallowBorrow,
+            ) => {}
+            _ => {
+                return;
+            }
         }
-        self.super_rvalue(rvalue, location);
-    }
 
-    fn visit_place(&mut self, place: &Place<'tcx>, context: PlaceContext, _location: Location) {
-        if let PlaceContext::NonUse(_) = context {
-            return;
-        }
         if !place.is_indirect() {
             return;
         }
 
-        let pointer = Place::from(place.local);
-        let pointer_ty = pointer.ty(&*self.local_decls, self.tcx).ty;
-
-        // We only want to check unsafe pointers
-        if !pointer_ty.is_unsafe_ptr() {
-            trace!("Indirect, but not an unsafe ptr, not checking {:?}", pointer_ty);
+        // We only want to check places based on unsafe pointers
+        let local_ty = self.local_decls[place.local].ty;
+        if !local_ty.is_unsafe_ptr() {
+            trace!("Indirect, but not based on an unsafe ptr, not checking {:?}", place);
             return;
         }
 
-        let Some(pointee) = pointer_ty.builtin_deref(true) else {
-            debug!("Indirect but no builtin deref: {:?}", pointer_ty);
-            return;
-        };
-        let mut pointee_ty = pointee.ty;
-        if pointee_ty.is_array() || pointee_ty.is_slice() || pointee_ty.is_str() {
-            pointee_ty = pointee_ty.sequence_element_type(self.tcx);
-        }
+        let place_ty = place.ty(self.local_decls, self.tcx).ty;
 
-        if !pointee_ty.is_sized(self.tcx, self.tcx.param_env_reveal_all_normalized(self.def_id)) {
-            debug!("Unsafe pointer, but unsized: {:?}", pointer_ty);
+        // The AlignOf MIR operator rejects unsized types
+        let pointee_ty = self.local_decls[place.local].ty.builtin_deref(true).unwrap().ty;
+        // FIXME: This ought to use naive layout, when that lands
+        if !pointee_ty.is_sized(self.tcx, self.param_env) {
+            debug!("Unsafe pointer, but unsized: {:?}", place_ty);
             return;
         }
+        // And also check the sources for all trailing field projections, we will need the
+        // alignment of those types later.
+        let mut temp_place = *place;
+        while matches!(temp_place.projection.as_slice().last(), Some(ProjectionElem::Field(..))) {
+            temp_place = Place::from(place.local).project_deeper(
+                &temp_place.projection[..temp_place.projection.len() - 1],
+                self.tcx,
+            );
```

Comment:
Can't this use the usual `PlaceRef` abstraction? Directly accessing `projections` and messing about with slices should be considered an anti-pattern.